Single-language, single-tab app translations upload

orangejenny opened the pull request:
[JIRA ticket](https://dimagi-dev.atlassian.net/browse/ICDS-295)

[QA ticket](https://dimagi-dev.atlassian.net/browse/QA-264)

[Original spec](https://docs.google.com/document/d/1ZAmUE7bHRmsNx6EYWcuE4pmu1eAm4LwDEkvb_8_Rzak/), although it has deviated a bit. The general idea is to make the download added in https://github.com/dimagi/commcare-hq/pull/23391 be able to be re-uploaded.

The UI has changed: this isn't really a multimedia-specific tool, even though ICDS is planning to only use it to update audio files, so it's now a part of the bulk app trans section instead of being its own section. If you have the flag turned on, you see the language dropdown, and if you pick a language, you get the single-tab file.

<img width="984" alt="Screen Shot 2019-03-14 at 8 30 40 PM" src="https://user-images.githubusercontent.com/1486591/54400192-10a91300-4698-11e9-972c-0a9620902a9f.png">

Review by commit. Many of the commits just extract helper functions and split `app_translations.py` into multiple files. It might be helpful to read the [README](https://github.com/dimagi/commcare-hq/blob/e3ca07e8e80728e80e959a3b6be3fb8e8f7ae9b9/corehq/apps/translations/app_translations/README.md) before reviewing.

Product: feature flag is "Bulk multimedia path management." This does make also minor changes to the all-users-facing bulk app translation, like reordering columns and changing the names of the sheets. All of these changes should be backwards compatible, so uploading a pre-release file should work. I'll remove the backwards compatibility a few days after release.

orangejenny committed:
Removed unused imports

This also reorders the columns in the bulk app translations so that
both the first modules+forms sheet and the individual form sheets
order media types the same way.

orangejenny committed:
Added skeletal upload view

orangejenny committed:
Extracted _cannot_add_sheet_error

orangejenny committed:
Extracted _check_for_sheet_warnings

orangejenny committed:
Extracted is_foo_sheet methods

orangejenny committed:
Renamed update_app_from_foo methods for consistency

orangejenny committed:
Extracted _get_condensed_rows

orangejenny committed:
Extracted _check_for_detail_length_errors

I don't love that this returns a tuple but still think it's an improvement.

orangejenny committed:
Extracted _check_for_shadow_form_warning

orangejenny committed:
Renamed app_translations.py in preparation for splitting it up

orangejenny committed:
Extracted functions into upload_app.py

orangejenny committed:
Extracted functions into upload_module.py

orangejenny committed:
Extracted functions into upload_form.py

orangejenny committed:
Extracted functions into utils.py

Also moved a couple of functions here that I had previously put in translations.utils.py
instead of translations.app_translations.utils.py

orangejenny committed:
Moved remaining functions to download.py

orangejenny committed:
Combined multi-sheet and single-sheet processing

Looks like it'll make sense to have the same code handle both the
existing app translations upload and the new single sheet upload.

This commit DRYs up a bunch of code but doesn't yet process
translations for the new upload.

orangejenny committed:
Combined multi-sheet and single-sheet upload in UI

orangejenny committed:
Added support for processing single-tab upload

orangejenny committed:
Renamed foo_filepath_lang headers to be consistent with other sheets.

Laurence requested. Will do a followup PR to remove the backwards compatibility code.

orangejenny committed:
Split detail_or_label column into list_or_detail and label columns.

orangejenny committed:
Added README

orangejenny committed:
Added friendlier error message if mixed up the two types of uploads

orangejenny committed:
Removed validation checkbox when dealing with single-sheet format

orangejenny committed:
Renamed functions that return dicts of sheets, not lists of rows

orangejenny committed:
Renamed 'menu' to 'module' in user-facing places

QA bug from the previous PR
This should be backwards compatible; I'll do another PR to revert that logic.

orangejenny committed:
Fixed import from bad rebase

czue commented on corehq/apps/translations/views.py:
```diff
@@ -118,6 +119,30 @@ def download_bulk_multimedia_translations(request, domain, app_id):
+    lang = request.POST.get('language')
+    app = get_app(domain, app_id)
+    workbook, msgs = get_app_translation_workbook(request.file)
+    if workbook:
```
when is workbook empty?

orangejenny replied:
I don't know - this code is pre-existing.

czue commented on corehq/apps/translations/app_translations.py:
```diff
@@ -95,60 +93,15 @@ def process_bulk_app_translation_upload(app, workbook):
-                'occurrence has been processed' %
-                sheet.worksheet.title
-            ))
+        error = _check_for_sheet_error(sheet, expected_sheets=expected_sheets, processed_sheets=processed_sheets)
```
this is a great change. one thought is that now that it's extracted to its own function you could manage control flow with exceptions which might be a little more explicit than relying on the return value being `None`

e.g. 
```python
try:
  _check_for_sheet_error(...)
except InvalidSheetException as e:
  msgs.append(e.message)
```

orangejenny replied:
810cb075d2f

orangejenny replied:
also 72ca8681493

czue commented on corehq/apps/translations/app_translations.py:
```diff
@@ -785,6 +794,7 @@ def has_translation(row_, langs):
                   " for the label '%s' in sheet '%s'") % (label, sheet.worksheet.title)
             ))
     # Update the translations
+    missing_cols = _get_missing_cols(app, sheet)
```
nit: comment should probably be under this line

orangejenny replied:
(resolved by later commit)

czue commented on corehq/apps/translations/app_translations.py:
```diff
@@ -650,6 +646,7 @@ def update_form_translations(sheet, rows, app):
     module_index = int(mod_text.replace("module", "")) - 1
     form_index = int(form_text.replace("form", "")) - 1
     form = app.get_module(module_index).get_form(form_index)
+    rows = get_unicode_dicts(sheet)  # fetch once, because sheet.__iter__ can only be called once
```
curious how this still works since it seems like `get_unicode_dicts(sheet)` now gets called multiple times?

orangejenny replied:
Not following this question - `get_unicode_dicts` is only called once on each sheet.

czue replied:
> `get_unicode_dicts` is only called once on each sheet

This answers the question - didn't realize that.

czue commented on corehq/apps/translations/app_translations.py:
```diff
@@ -1108,6 +1031,95 @@ def _partial_upload(rows, details):
+    This function also pulls out case detail tab headers and the case list form label,
+    which will be processed separately from the case proeprty rows.
+
+    Returns a three-item tuple: case property rows, tab header rows, case list form label
```
could make this a `namedtuple` (is very low overhead and a just bit more explicit)

orangejenny replied:
bc4b19e8672

czue commented on corehq/apps/translations/app_translations.py:
```diff
@@ -898,39 +898,9 @@ def update_app_from_module_sheet(app, sheet):
-
-    if msgs:
-        return msgs
+    (errors, partial_upload) = _check_for_detail_length_errors(module, condensed_rows)
```
same comment (especially re: commit message) about `namedtuple` could apply here

orangejenny replied:
bc4b19e8672

czue commented on corehq/apps/translations/app_translations/upload_module.py:
```diff
@@ -0,0 +1,291 @@
+
+    :param sheet:
+    :param app:
+    :return:  Returns a list of message tuples. The first item in each tuple is
```
repeating myself a lot but could be `namedtuple` here too. though also realize this was just moved.

orangejenny replied:
Leaving this one alone.

czue reviewed (commented):
Looks great overall and like a really nice cleanup. I had a few small non-blocking suggestions and one question.

What is our test coverage for these different pathways and uploads? It would be great if we could get reasonable test coverage for the different formats and error cases.

kaapstorm commented on corehq/apps/translations/app_translations/upload_module.py:
```diff
@@ -0,0 +1,291 @@
+    """
+    Returns true if the given row has at least one translation.
+
+    >> has_at_least_one_translation(
```
I know you didn't write this docstring, but is it intentionally not a doctest? 

If you change `>>` to `>>>`, `true` to `True` and `false` to `False`, then you can get the test suite to [check the docstring](https://github.com/dimagi/commcare-hq/blob/868436c0a4c1a4daaeefd6c89ef6310af3d542e7/corehq/sql_db/tests/test_backports.py#L50) is correct with 

```python
def test_doctests():
    results = doctest.testmod(upload_module)
    assert results.failed == 0
```

orangejenny replied:
902c98d4013

kaapstorm commented on corehq/apps/translations/app_translations/README.md:
```diff
@@ -0,0 +1,21 @@
+
+The UI defaults to the multi-sheet download. There's a feature flagged ability to select a single language, which downloads the single sheet format for that language. There's no technical reason that the single sheet download is also a single language download.
+
+For domains with Transifex integration, Transifex generates generates excel files in the multi-sheet format that HQ accepts.
```
`generates generates`; nit: "Excel" is a proper noun.

kaapstorm commented on corehq/apps/translations/app_translations/README.md:
```diff
@@ -0,0 +1,21 @@
+For domains with Transifex integration, Transifex generates generates excel files in the multi-sheet format that HQ accepts.
+
+Code is organized into
+- `download.py` Generation of excel downloads
```
nitto

orangejenny replied:
bcaf56b2d00

kaapstorm commented on corehq/apps/translations/app_translations/upload_app.py:
```diff
@@ -0,0 +1,284 @@
+    :param app:
+    :return:  Returns a list of message tuples. The first item in each tuple is
+    a function like django.contrib.messages.error, and the second is a string.
+    """
```
nit: Docstring no longer matches function signature.

orangejenny replied:
d6938292152

kaapstorm reviewed (approved):
I love the refactors.

Commit:
Merge branch 'master' into jls/upload-mm-translations

mkangia commented on corehq/apps/translations/app_translations.py:
```diff
@@ -196,6 +145,30 @@ def process_bulk_app_translation_upload(app, workbook):
     return msgs
 
 
+def _check_for_sheet_error(sheet, expected_sheets=Ellipsis, processed_sheets=Ellipsis):
```
i'd just call this `validate_sheets` or `validate_workbook`.

orangejenny replied:
I like the symmetry of `_check_for_workbook_error`, `_check_for_sheet_error`, and `_check_for_sheet_warnings`

mkangia commented on corehq/apps/translations/app_translations.py:
```diff
@@ -862,7 +862,7 @@ def has_translation(row_, langs):
     try:
         return etree.fromstring("<value>{}</value>".format(
-            re.sub(r"(?<!/)>", "&gt;", re.sub(r"<(\s*)(?!output)", "&lt;\\1", value))
+            re.sub("(?<!/)>", "&gt;", re.sub("<(\s*)(?!output)", "&lt;\\1", value))
```
why this change?

orangejenny replied:
Not intentional, I must have done a bad rebase incorporating https://github.com/dimagi/commcare-hq/pull/23403

Fixing in 7150aa0e9fc

mkangia replied:
you missed the second "r" in the fix.

mkangia commented on corehq/apps/translations/app_translations/README.md:
```diff
@@ -0,0 +1,21 @@
+Bulk Application Translations
```
👍

mkangia commented on corehq/apps/translations/app_translations/upload_app.py:
```diff
@@ -0,0 +1,284 @@
+
+    expected_headers = _get_expected_headers(sheet, expected_sheets)
+    if expected_headers is None:
+        return _('Skipping sheet "%s", did not recognize title') % sheet.worksheet.title
```
nit: `could` not recognize?

orangejenny replied:
d32e3694a2e

mkangia commented on corehq/apps/translations/app_translations/upload_form.py:
```diff
@@ -0,0 +1,278 @@
+
+
+
+def update_app_from_form_sheet(app, rows, identifier):
```
This is one method :(

orangejenny replied:
Split up in a couple of commits: d3885475dc049f74bc1fa31177194ceab059b004 b589442e80342e01722bd18257549456fea28566 251a6320b43a65467bcc4ab855470b5427e9f14b 8e58c1567f6ae6cd1e1de0a6c658a0ef7222f40a

mkangia replied:
Thank you :)

mkangia reviewed (commented):
Do you plan to break the longer methods in near future @orangejenny ?

This is definitely better with better collections of methods.. though we still have very long methods.

orangejenny committed:
Fixed ilnt errors

orangejenny committed:
Merge branch 'jls/upload-mm-translations' of github.com:dimagi/commcare-hq into jls/upload-mm-translations

orangejenny committed:
Replaced a few tuples with namedtuples

orangejenny committed:
Fixed grammar in README

orangejenny committed:
Updated docstring

orangejenny committed:
Added doctest

kaapstorm commented:
Thanks @orangejenny

orangejenny committed:
Updated _check_for_sheet_error to raise exception instead of returning error

orangejenny committed:
Fixed regex

orangejenny committed:
Updated error wording and fixed a few lint issues

orangejenny committed:
Fixed indentation error

orangejenny committed:
Merge branch 'master' into jls/upload-mm-translations

orangejenny committed:
Fixed 500 for shadow forms and report modules (w=1)

orangejenny committed:
Updated _check_for_shadow_form_error to raise exception instead of returning error

orangejenny committed:
Moved inner functions out of update_app_from_form_sheet

Now that this file is just this one function, it's more readable
to have the helper functions after update_app_from_form_sheet rather
than in the middle of it.

orangejenny committed:
Extracted more helpers out of update_app_from_form_sheet

mkangia commented:
Build failures look related
```
======================================================================
ERROR: corehq.apps.translations.tests.test_bulk_app_translation:BulkAppTranslationBasicTest.test_bad_column_name
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/mnt/commcare-hq-ro/corehq/apps/translations/tests/test_bulk_app_translation.py", line 411, in test_bad_column_name
    'App Translations Updated!'
  File "/mnt/commcare-hq-ro/corehq/apps/translations/tests/test_bulk_app_translation.py", line 80, in upload_raw_excel_translations
    messages = process_bulk_app_translation_upload(self.app, workbook, expected_headers)
  File "/mnt/commcare-hq-ro/corehq/apps/translations/app_translations/upload_app.py", line 102, in process_bulk_app_translation_upload
    sheet_name=sheet.worksheet.title))
  File "/mnt/commcare-hq-ro/corehq/apps/translations/app_translations/upload_app.py", line 121, in _process_multi_sheet_rows
    return update_app_from_form_sheet(app, rows, identifier)
  File "/mnt/commcare-hq-ro/corehq/apps/translations/app_translations/upload_form.py", line 152, in update_app_from_form_sheet
    {'form': trans_type})
  File "/mnt/commcare-hq-ro/corehq/apps/translations/app_translations/upload_form.py", line 200, in _update_translation_node
    text_node.xml.append(e)
NameError: global name 'text_node' is not defined
----------------------------------------------------------------------
```

```
======================================================================
ERROR: corehq.apps.translations.tests.test_bulk_app_translation:AggregateMarkdownNodeTests.test_markdown_node
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/mnt/commcare-hq-ro/corehq/apps/translations/tests/test_bulk_app_translation.py", line 803, in test_markdown_node
    msgs = update_app_from_form_sheet(self.app, sheet, sheet.worksheet.title)
  File "/mnt/commcare-hq-ro/corehq/apps/translations/app_translations/upload_form.py", line 140, in update_app_from_form_sheet
    delete_node=(not keep_value_node)
  File "/mnt/commcare-hq-ro/corehq/apps/translations/app_translations/upload_form.py", line 200, in _update_translation_node
    text_node.xml.append(e)
NameError: global name 'text_node' is not defined
----------------------------------------------------------------------
```

kaapstorm commented on corehq/apps/translations/app_translations/upload_form.py:
```diff
@@ -255,7 +255,7 @@ def _check_for_shadow_form_warning(form):
     try:
         return etree.fromstring("<value>{}</value>".format(
-            re.sub("(?<!/)>", "&gt;", re.sub("<(\s*)(?!output)", "&lt;\\1", value))
+            re.sub(r"(?<!/)>", "&gt;", re.sub("<(\s*)(?!output)", "&lt;\\1", value))
```
I was curious what you were going to do with @mkangia's comment, @orangejenny, because I thought dropping the `r` prefixes was fine. :grin:

@mkangia you probably know this already, but just to confirm we're on the same page, the `r` prefix just means "[all backslashes are left in the string](https://docs.python.org/2.0/ref/strings.html)". The first pattern has no backslashes, so the `r` has no effect, and the second pattern has one, but `\s` is not an escape sequence, so the `r` has no effect there either.

... I was going to leave this comment at that, but I went looking for what Python 3 docs said about raw string literals, and kept coming across regex patterns that had `r` prefixes even though they clearly didn't need them because they didn't have any backslashes. [Then I found](https://docs.python.org/3/library/re.html#regular-expression-syntax) "... it’s highly recommended that you use raw strings for all but the simplest expressions". Hmmm. Maybe this is a convention that GitHub's syntax highlighting assumes and I was just unaware of. (In my defence, PyCharm uses regex syntax highlighting for all patterns, regardless of whether they are raw string literals or not.)

Anyway, fwiw, I'm fine with these patterns as they are. I'm also fine without the `r` prefix because they don't need them. But if using raw strings for all regex patterns is a thing I didn't know about, we should probably just always use them.

mkangia replied:
Thanks for sharing that @kaapstorm. I just choose to not modify what i don't understand completely or whose intent is not clear to me. After reading your note i confused myself with somethings i tried with re.sub, so i am going to leave this to Jenny to decide to change or not.

orangejenny replied:
Based on https://github.com/dimagi/commcare-hq/pull/23403 my understanding is that our convention is to use the `r` prefix for all regular expressions. @nickpell is that accurate?

orangejenny committed:
Extracted final helper functions from update_app_from_form_sheet

orangejenny committed:
Fixed lint errors / bugs

orangejenny committed:
Merge branch 'master' into jls/upload-mm-translations

orangejenny commented:
All code review feedback should now be addressed - thank you for all of the feedback.

@mkangia I'm not planning to do further refactoring, but let me know if there's anything that's still particularly egregious and I'll consider it ;)

@czue Good point about test coverage. It's relatively good in this area but could certainly be better, especially since I'm adding new code. I'll take a pass at adding tests.

mkangia reviewed (approved):
Thank you for stretching the scope of this PR for code improvement.

czue commented:
thanks @orangejenny !

orangejenny committed:
Added error message to ModuleNotFoundException

orangejenny committed:
Merge branch 'master' into jls/upload-mm-translations

orangejenny committed:
Updated translation logic to only look at selected language

Also moved some inner functions to the module level.

orangejenny committed:
Replaced helper functions in upload_form.py with a class (w=1)

orangejenny committed:
Fixed lint errors

orangejenny committed:
Replaced helper functions in upload_module.py with a class (w=1)

orangejenny committed:
Replaced sheet_name header in multi-sheet download with menu_or_form.

Should be backwards compatible; will remove this logic in a future PR.

orangejenny committed:
Added BulkAppTranslationModulesAndFormsUpdater (w=1)

orangejenny committed:
Moved update_translation_dict into BulkAppTranslationUpdater

orangejenny committed:
Fixed lint errors

orangejenny commented:
Added a few more refactoring commits, grouping some related helper functions into classes to reduce the amount of state they pass around.

Still working on increasing test coverage.

orangejenny committed:
Increased feature coverage of BulkAppTranslationDownloadTest

Adds case detail tabs, graphing, ID mapping case property.

orangejenny committed:
Merge branch 'master' into jls/upload-mm-translations

orangejenny committed:
Reconciled language selection UI for validation and download/upload.

This removes most of the UI code added in https://github.com/dimagi/commcare-hq/pull/23817/
and instead uses the same language dropdown for download/upload and validation.

orangejenny committed:
Added tests for some worksheet error conditions.

And fixed some bugs with formulating those error messages.

orangejenny committed:
Added tests for sheet-level errors

orangejenny committed:
Fixed test class inheritance.

Fixup for b5198a93f50a765889ec970ab5532284920cebff

orangejenny committed:
Added remainder of test data for new feature coverage

Fixup for a274ec00e59805c4d0e7b84996d477da605afd67

orangejenny committed:
Fixed lint errors

orangejenny committed:
Added upload test for single-sheet format

orangejenny commented:
@czue Added test coverage.

czue commented:
> @czue Added test coverage.

Hooray, thank you!

orangejenny committed:
Merge branch 'master' into jls/upload-mm-translations

orangejenny committed:
Added logic to hide and ignore validation checkbox unless lang is specified

orangejenny committed:
Added better backwards compatibility for old "sheet_name" header

The upload was working, but it was still displaying warnings about
unrecognized and unexpected columns.

emord commented on corehq/apps/translations/app_translations/README.md:
```diff
@@ -0,0 +1,21 @@
+
+The download has two variants, a multi-sheet and a single-sheet format. Both are tolerant of partial uploads.
+
+The default multi-sheet format contains a first "menus and forms" sheet for editing module and form names and menu media. It then contains a sheet for each module and each form.
```
should the modules in this line be called menus?

orangejenny replied:
Eh, the code primarily uses "module" so I think it's fine if the docs do.

emord commented on corehq/apps/translations/integrations/transifex/parser.py:
```diff
@@ -13,7 +13,7 @@
 CONTEXT_REGEXS = {
     # Module or Form: sheet name for module/form: unique id
-    'module_and_forms_sheet': r'^(Module|Form):(\w+):(\w+)$',
+    'module_and_forms_sheet': r'^(Menu|Form):(\w+):(\w+)$',
```
should the other instances of module be changed to menu as well

orangejenny replied:
Are there particular ones you noticed? It's a bit hairy in this file because the transifex resources are, and should be, still named "module" - I don't want to be changing the names of anything in transifex. I'm fairly dependent on @mkangia to review these changes.

emord reviewed (commented):
Did a quick pass over this. I'm not quite sure where we're drawing the line between module and menus here

I think I may be missing it, but where are the xlsx files in app_manager being used? All of the tests I'm seeing use python objects to upload translations

orangejenny committed:
Updated docstring

orangejenny commented on corehq/apps/translations/app_translations/utils.py:
```diff
@@ -0,0 +1,226 @@
+    This function is meant as the single point of truth for the
+    column ordering of {sheet_name}
+
+    """.format(sheet_name=MODULES_AND_FORMS_SHEET_NAME)
```
Don't know what was up with that, don't think I added it. 9c494dd7b72

orangejenny commented on corehq/apps/app_manager/tests/data/bulk_app_translation/download/app.json:
```diff
@@ -394,10 +394,1913 @@
         }
       }
+    },
+    {
```
Unfortunately not - this change makes the download test pretty comprehensive, but many of the features being tested (e.g., graphs) aren't supported by AppFactory.

orangejenny committed:
Updated README

orangejenny commented on corehq/apps/translations/app_translations/README.md:
```diff
@@ -0,0 +1,21 @@
+
+HQ supports a file download and re-upload to update all application-specific translations.
+
+The download has two variants, a multi-sheet and a single-sheet format. Both are tolerant of partial uploads.
```
240f7ac54ca

> From what I understand, only a subset of languages is supported without a feature flag

What do you mean? I think you can leave out rows. Maybe see if the update I made sounds accurate to you?

emord replied:
I was thinking of this PR https://github.com/dimagi/commcare-hq/pull/23367

However maybe that doesn't apply here as those are "UI" translations and not in the app_translations folder?

orangejenny replied:
Yeah, that flag is only CommCare translations, not app translations.

orangejenny commented:
> I'm not quite sure where we're drawing the line between module and menus here

Everything user-facing in HQ should be menu; all of the code should be module. I don't know if the transifex resource names are user-facing in the transifex interface, but if they are, I don't care that they still say module - the "module" => "menu" change is more for SaaS users.

> I think I may be missing it, but where are the xlsx files in app_manager being used? All of the tests I'm seeing use python objects to upload translations

https://github.com/dimagi/commcare-hq/blob/ffd94d714fff2482f304afb85cafcf213e0f644c/corehq/apps/translations/tests/test_bulk_app_translation.py#L86

emord commented:
> Everything user-facing in HQ should be menu; all of the code should be module

Is there a technical reason for this other than changing it would be difficult? (not trying to sound judge-y, just want to understand where the separation is and the reason for it. I realize that change would be a slog and prone to errors)

mkangia commented on corehq/apps/translations/generators.py:
```diff
@@ -250,14 +250,14 @@ def _get_translation_for_sheet(self, app, sheet_name, rows):
             unique_id_index = self._get_header_index(MODULES_AND_FORMS_SHEET_NAME, 'unique_id')
 
             def occurrence(_row):
                 return ':'.join([_row[type_index], _row[sheet_name_index], _row[unique_id_index]])
```
This line currently generates this `Module:module1:ad7059becbdcd3dfe155c116d5ac6837017b4e31` but after this change it would generate `Menu:menu1:ad7059becbdcd3dfe155c116d5ac6837017b4e31`. We would want to keep it `Module` and `module` only because a change in occurrence would cause the existing translations to get removed.

orangejenny replied:
thank you - https://github.com/dimagi/commcare-hq/pull/23924

orangejenny committed:
Merge branch 'master' into jls/upload-mm-translations

orangejenny committed:
Moved test data files from app_manager to translations

orangejenny committed:
Renamed upload_no_change_data and removed duplicate headers list

orangejenny committed:
Added failing test

orangejenny committed:
Fixed bug

orangejenny committed:
Added failed text for node deletion bug

orangejenny committed:
Fixed bug with deleting nodes when only uploading single sheet

orangejenny committed:
Removed fallback logic if there's only a single language

orangejenny committed:
Moved condition; committed too soon

orangejenny committed:
Merge pull request #23912 from dimagi/jls/move-test-data-files

Moved test data files from app_manager to translations

orangejenny committed:
Fixed sheet_name_index in transifex

orangejenny commented:
Doing all further updates in separate PRs since this comment thread is so long. I'll continue to merge master into this branch; rebasing has gotten hairy.

orangejenny committed:
Merge branch 'master' into jls/upload-mm-translations

orangejenny committed:
Added is_multi_sheet flag

orangejenny committed:
Fixed test data.

In the multi-sheet upload, there's a value for item2, so there should be a value
for item2 in the xml in both english (provided) and french (fallback).

mkangia commented on corehq/apps/hqwebapp/forms.py:
```diff
@@ -111,34 +107,23 @@ def crispy_form_fields(self, context):
+        if context.get('can_select_language'):
+            crispy_form_fields.extend([
+                InlineField('language', data_bind="value: lang")
+            ])
```
we would need to change this to also have the language field added in case `can_validate_app_translations` is true to support https://github.com/dimagi/commcare-hq/pull/23817 ?

mkangia commented on corehq/apps/hqwebapp/forms.py:
```diff
@@ -111,34 +107,23 @@ def crispy_form_fields(self, context):
+                )
             ])
-            self.fields['language'].choices = tuple_of_copies(self.app.langs, blank=False)
-            self.fields['language'].initial = self.app.default_language
```
Looks like the choices is still all app langs but the initial value is "All" instead of the default language as set up for the validation feature. Would need to address that for the validation feature, I can take that in a follow up PR once this is merged.

orangejenny replied:
The UI does now default to "All" and hides the validation checkbox. The validation shows up when you pick a language: http://g.recordit.co/3VbpJ6Z4nG.gif

I'm not totally in love with this UX, but we shouldn't have more than one language dropdown in the same form. If this UI is confusing, we should look at separating validation from uploading, rather than having validation be a checkbox.

mkangia replied:
Interesting work around there. Would just need to make people aware of this change, i am sure they would be searching for the validation option.

mkangia commented on corehq/apps/translations/views.py:
```diff
@@ -94,48 +92,36 @@ def download_bulk_ui_translations(request, domain, app_id):
     if workbook:
         if validate:
-            msgs = validate_bulk_app_translation_upload(app, workbook, request.user.email, request.file,
-                                                        request.POST.get('language'))
```
i think you removed request.file during master update. It should be added and later run in `validate_bulk_app_translation_upload` if you look at the current state of the method.

orangejenny replied:
Are you sure? The way I read the code, `request.file` gets turned into `workbook` a few lines up, and then `workbook` is how the data is passed to `validate_bulk_app_translation_upload`.

mkangia commented on corehq/apps/translations/app_translations/utils.py:
```diff
@@ -0,0 +1,226 @@
+        MODULES_AND_FORMS_SHEET_NAME,
+        get_modules_and_forms_row(
+            row_type='Type',
+            sheet_name='menu_or_form',
```
trying to convince you for a better name here if possible.
menu_or_form would have values like `module1` and `module1_form2` right?
`menu_or_form` sounds like this would have values ['module', 'form']. I would push to rename this back to `sheet_name` only or `module_or_form_name` instead.

orangejenny replied:
I can change to `module_or_form_name`. I'd rather not change back to `sheet_name` because that doesn't make sense for the single-sheet format, where these strings aren't sheet names.

orangejenny replied:
Rather, I'll change to `menu_or_form_name`.

orangejenny replied:
...although I just realized that `menu_or_form_name` sounds like that column should contain the name, not a string like `menu1`. It'd be more accurate to use `menu_or_form_slug`, but `slug` isn't a user-friendly name. So I agree that `menu_or_form` sounds like values would be either `menu` or `form` but I don't have a better name. Could do something really generic, like `item`, but that isn't really clear either.

mkangia committed:
use legacy notation to use module instead of menu

switching to new notation would cause the current translations to expire since the context would change

orangejenny committed:
Merge branch 'jls/upload-mm-translations' into jls/upload-mm-translations-fix-deletions

orangejenny committed:
Merge pull request #23915 from dimagi/jls/upload-mm-translations-fix-deletions

Fix deletion of non-uploaded languages

orangejenny committed:
Cast keys to list so we can pop from dict

orangejenny committed:
Merge pull request #23958 from dimagi/jls/staging-keys-to-list

Cast keys to list so we can pop from dict

orangejenny committed:
Added test that content in old languages gets cleared

orangejenny committed:
Merge pull request #23924 from dimagi/jls/upload-mm-translations-fix-transifex-module

Fixed sheet_name_index in transifex

orangejenny committed:
Merge pull request #23959 from dimagi/jls/add-content-clear-test

Added test that content in old languages gets cleared

Commit:
Merge branch 'master' into jls/upload-mm-translations

orangejenny committed:
Fixed spacing on error message
